libdnf5/transaction/transaction_replay.cpp

```diff
@@ -0,0 +1,425 @@
+                reason = transaction_item_reason_from_string(json_object_get_string(value));
+            }
+            if (json_object_object_get_ex(rpm, "group", &value) != 0) {
+                group = json_object_get_string(value);
+            }
+            if (json_object_object_get_ex(rpm, "repo_id", &value) != 0) {
+                repo_id = json_object_get_string(value);
+            }
+
+
+            // Find the required package
+            libdnf5::rpm::PackageQuery pkg_query(base);
+
+            // In case the package is found in the same repo as in the original transaction, limit the query to that
+            // plus installed packages. IOW remove packages with the same NEVRA in case they are found in multiple
+            // repos and the repo the package came from originally is one of them.
+            // This can e.g. make a difference in the system-upgrade plugin, in case the same NEVRA is in two repos,
+            // this makes sure the same repo is used for both download and upgrade steps of the plugin.
+            if (!repo_id.empty()) {
+                if (repo_id == "@System") {
+                    pkg_query = installed;
+                } else {
+                    libdnf5::rpm::PackageQuery query_repo(base);
+                    query_repo.filter_repo_id({repo_id});
+                    if (!query_repo.empty()) {
+                        pkg_query = query_repo;
+                        pkg_query |= installed;
+                    }
+                }
+            }
+
+            auto nevra_pair = pkg_query.resolve_pkg_spec(
+                nevra,
+                libdnf5::ResolveSpecSettings{.with_provides = false, .with_filenames = false, .with_binaries = false},
+                false);
+            if (!nevra_pair.first) {
+                raise_or_warn(skip_unavailable, fmt::format("Cannot find rpm nevra {:s}", nevra));
+                continue;
+            }
+            //TODO(amatej): verify package checksums
+
+            if (action == "Install" || action == "Upgrade" || action == "Downgrade" || action == "Reinstall") {
+                libdnf5::rpm::PackageQuery pkg_query_na(installed);
+                pkg_query_na.filter_name({nevra_pair.second.get_name()});
+                pkg_query_na.filter_arch({nevra_pair.second.get_arch()});
+                libdnf5::rpm::PackageQuery pkg_query_installonly(pkg_query_na);
+                pkg_query_installonly.filter_installonly();
+
+                if (action == "Install" && !pkg_query_na.empty() && pkg_query_installonly.empty()) {
+                    raise_or_warn(
+                        ignore_installed,
+                        fmt::format(
+                            "Package {:s}.{:s} is already installed, it cannot be installed again.",
+                            nevra_pair.second.get_name(),
+                            nevra_pair.second.get_arch()));
+                }
```

Comment:
I am quite curious why you run the check for `install` action. If I am not mistaken the request will be validate by libsolv including install-only rule. May I ask you for additional information?

Reply:
I have taken this behavior from the dnf4 replay however looking into it I think its there to have a working `--ignore-installed` option.
The option controls whether the command fails if we try to install already installed package.

I see that goal has the same check and it logs `GoalProblem::ALREADY_INSTALLED` but as far as I can see it is only logged it cannot fail the transaction.
I did some testing and it seems libsolv doesn't have a problem with installing already installed pkg, it either reinstalls, upgrades or downgrades the package based on the versions.